### ISSUE TITLE
different publish behavior between visual studio 2019 and 2022

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/runtime#58894]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/different-publish-behavior-between-visua/1518600)._

---
[regression] [worked-in:16.11]
Hello, i Just upgrade my visual studio 2019 to 2022 and check everything is working fine on my solution (.NET 5.0)

I notice a difference on the publish of one my app, especially the size of the publish app. before it was 320 elements for 80 Mo, now its 401 elements for 120 Mo. What can explain this difference ? i target the .NET 5.0.0 on my project.

this is the command i use

```
dotnet publish Is.Geckos.LocalService.csproj -p:PublishProfile=FolderProfile --configuration Release
```

the publish profile

```
<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <PropertyGroup>
    <WebPublishMethod>FileSystem</WebPublishMethod>
    <LastUsedBuildConfiguration>Release</LastUsedBuildConfiguration>
    <LastUsedPlatform>Any CPU</LastUsedPlatform>
    <SiteUrlToLaunchAfterPublish />
    <LaunchSiteAfterPublish>True</LaunchSiteAfterPublish>
    <ExcludeApp_Data>False</ExcludeApp_Data>
    <ProjectGuid>a9a4c861-924a-4190-bc23-fd99cf48edd2</ProjectGuid>
    <publishUrl>bin\Release\net5.0-windows\publish\</publishUrl>
    <DeleteExistingFiles>True</DeleteExistingFiles>
    <TargetFramework>net5.0-windows</TargetFramework>
    <RuntimeIdentifier>win-x86</RuntimeIdentifier>
    <SelfContained>true</SelfContained>
    <PublishTrimmed>True</PublishTrimmed>
  </PropertyGroup>
</Project>
```

i notice this warning during the publish which was not present before (im curious why the .NET 6 is invoke here, is it the reason why app is bigger)

```
C:\Program Files\dotnet\sdk\6.0.100-preview.7.21379.14\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(168,5): warning NETSDK1175: Windows Forms is not supported or recommended with trimming enabled. Please go to https://aka.ms/dotnet-illink/windows-forms for more details. 

```

other wise the app work well, and i can confirm she is on .. NET 5. but it could be nice to come back at initial size because it make an augmentation of 50% !

thanks !

---
#### Feedback Bot on 8/31/2021, 05:55 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by vitek-karas** [comment]
I have to start with what the warning says: Currently, trimming WinForms apps is unsupported and it's known to break apps frequently.

By default VS will use the latest SDK version, and VS2022 installs .NET 6 previews - so it's used. Normally this should not matter, since SDK should be backward compatible. But in .NET 5 trimming was a preview feature, so we don't maintain the same backward compatibility standards as with other functionality. In .NET 6 we improved lot of things around trimming, and the behavior changed quite a bit.

That said I've not seen such a big difference yet. It would be interesting to know where the difference comes from. (a simple WinForms app produces only a small (~100KB) difference for me)

Aside from that, you can tell .NET to keep using .NET 5 SDK for your project. Add a [`global.json`](https://docs.microsoft.com/en-us/dotnet/core/tools/global-json?tabs=netcore3x) file to your project root and specify something like:
```json
{
  "sdk": {
    "version": "5.0.400"
  }
}
```

This will make sure that the project will be built with the specified version of SDK, regardless of what else is installed on the machine. And that means it should behave the same as before.

**Comment by julienGrd** [comment]
my app is self contained, thats'explain this size. for example, with my old environnment, the app was publish with the dll PresentationFamework.dll of 5781Ko and with my new installation its 14450Ko. So it seem indeed the trimming is disabled for this part.

But even with putting a global.json (and also a <RuntimeFrameworkVersion>5.0.0</RuntimeFrameworkVersion> in the csproj), i have the same result.

My app actually work well in .net 5 with trimming enabled, is it possible to lert this functionnality work but warn the user is it "at his own risk ?

**Comment by vitek-karas** [comment]
The global.json should work. You can try this from command line. Run `dotnet --version` from the project directory. It should print out the effective SDK version - which should be 5.0.x. If it's not, the global.json is probably not right. I think the default behavior is exact version match - which is somewhat invonvenient. You can try this instead:

```json
{
  "sdk": {
    "version": "5.0.0",
    "rollForward": "Minor"
  }
}
```

**Comment by vitek-karas** [comment]
One more question - what is the `TargetFramework` of your app? The warning about winforms not supporting trimming and the fact that the output is much bigger would make sense if the app is `net6.0`, but it should not behave that way for `net5.0` apps.

**Comment by agocke** [comment]
@vitek-karas The block on WinForms and WPF is not conditional on TFM https://github.com/dotnet/sdk/blob/fb573e46f9f78e2213d1f1b580992a518f7c1589/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets#L176

**Comment by vitek-karas** [comment]
Sorry - I probably did something wrong. I tried again and I see the same size difference. I'm still looking into why is that - but the fix with global.json is still valid (and should be used in this case anyway).

**Comment by vitek-karas** [comment]
The difference is R2R data. In 5.0 the trimming had a very confusing behavior where it sometimes kept the assembly intact and sometimes it would "save" it from memory. Saving from memory removes R2R data. In 6.0 we made this behavior much more consistent and the trimmer now leaves the assembly intact in more cases (as it's the correct thing to do). This leads to larger output size.

After offline discussion the proposal to handle WinForms in 6.0 is:
* SDK will actually fail to trim a WinForms app (regardless of TFM) - the warning shown above is an error (already true in RC1 + builds)
* We will update the page that error points to to include section about WinForms specifically and describe the workaround of using global.json to stay on 5.0 SDK if needed.

Reasoning:
* Trimming in 5.0 was a preview feature - as such there's no guarantee to its (sometimes wrong) behavior.
* Trimming in 6.0 is not a preview feature - as such it tries to be much more precise in providing feedback on what will and will not work. We know that trimming WinForms doesn't consistently work
* If we were to make net5.0 apps work more or less the same as they did even in 6.0 SDK we would have to mostly duplicate often time incorrect or incomplete behavior from the preview feature. This seems wrong and too expensive to do.

**Comment by julienGrd** [comment]
dotnet --version on the project give 5.0.100
the targetFramework is set as net5.0-windows, OutputType WinExe, UseWindowsForms at true

I don't know if you need more information, fo now i still have this behavior

**Comment by vitek-karas** [comment]
Just to make this clear:
* You added `global.json` and the result is that running `dotnet --version` from the project's directory outputs `5.0.100`
* When you publish the project with trimming you get a warning like:
  ```
  C:\Program Files\dotnet\sdk\6.0.100-preview.7.21379.14\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(168,5): warning NETSDK1175: Windows Forms is not supported or recommended with trimming enabled. Please go to https://aka.ms/dotnet-illink/windows-forms for more details. 
  ```
* The output is much larger than before

Can you please try deleting the `obj` and `bin` subdirectories and publishing again?

**Comment by julienGrd** [comment]
> Just to make this clear:
> 
> * You added `global.json` and the result is that running `dotnet --version` from the project's directory outputs `5.0.100`
> * When you publish the project with trimming you get a warning like:
>   ```
>   C:\Program Files\dotnet\sdk\6.0.100-preview.7.21379.14\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(168,5): warning NETSDK1175: Windows Forms is not supported or recommended with trimming enabled. Please go to https://aka.ms/dotnet-illink/windows-forms for more details. 
>   ```
> * The output is much larger than before
> 
> Can you please try deleting the `obj` and `bin` subdirectories and publishing again?

yes exactly. i deleted the bin and obj and publish again, il have the same results. i also confirm the published app is ont .net 5.0 (when i make envirnonment version). it only seem its the trimming which is disabled.

concerning the C:\Program Files\dotnet\sdk\6.0.100-preview.7.21379.14\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets, its the only line of the output which make reference to .NET6. otherwise its only .NET 5 reference 
like
````
 -a "C:\Users\julie\.nuget\packages\microsoft.windowsdesktop.app.runtime.win-x86\5.0.0\runtimes\win-x86\lib\net5.0\PresentationFramework.dll"
-reference "C:\Users\julie\.nuget\packages\microsoft.netcore.app.runtime.win-x86\5.0.0\runtimes\win-x86\lib\net5.0\mscorlib.dll"
````

**Comment by julienGrd** [comment]
new behavior this morning after upgrading to visual studio 2022 preview 4 (by extension to .net 6 RC1)

Its not oublish anymore, its finish with error 
C:\Program Files\dotnet\sdk\6.0.100-rc.1.21458.32\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(176,5): Erreur NETSDK1175: Windows Forms is not supported or recommended with trimming enabled. Please go to https://aka.ms/dotnet-illink/windows-forms for more details.

Whan i disable trimming for the publish, i have a 444 elements for 145 Mo

So its the double than the same app in .NET5

As i understand, you will not rollback on this for.NET 6. its really inconveniant, my company will never accept than  the size of the app double. so i think im blocked on .NET 5.

Anyway, its not the original problem, which is on my case the behavior should not produce because i target the .NET5.

**Comment by vitek-karas** [comment]
I tried again from a clean slate, to verify that the `global.json` works as expected. I honestly can't explain the behavior you're seeing. Below if what I see:

```
❯  dotnet --version
6.0.100-rc.1.21430.44

> mkdir winformstest

> cd winformstest

> dotnet new globaljson
The template "global.json file" was created successfully.

> notepad global.json

> type global.json
{
  "sdk": {
    "version": "5.0.401"
  }
}

❯  dotnet --version
5.0.401

❯  dotnet new winforms
The template "Windows Forms App" was created successfully.

Processing post-creation actions...
Running 'dotnet restore' on ...\winformstest\winformstest.csproj...
  Determining projects to restore...
  Restored ...\winformstest\winformstest.csproj (in 119 ms).
Restore succeeded.

❯  dotnet publish -r win-x64 -p:PublishTrimmed=true
Microsoft (R) Build Engine version 16.11.0+0538acc04 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restored ...\winformstest\winformstest.csproj (in 18.17 sec).
  winformstest -> ...\winformstest\bin\Debug\net5.0-windows\win-x64\winformstest.dll
  Optimizing assemblies for size, which may change the behavior of the app. Be sure to test after publishing. See: https://aka.ms/dotnet-illink
  winformstest -> ...\winformstest\bin\Debug\net5.0-windows\win-x64\publish\

```

Then I opened the project in VS and created a publish profile for win-x64 self-contained application with trimming enabled. This created `FolderProfiler.pubxml` like this:

```
❯  type .\Properties\PublishProfiles\FolderProfile.pubxml
<?xml version="1.0" encoding="utf-8"?>
<!--
https://go.microsoft.com/fwlink/?LinkID=208121.
-->
<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <PropertyGroup>
    <Configuration>Release</Configuration>
    <Platform>Any CPU</Platform>
    <PublishDir>bin\Release\net5.0-windows\publish\</PublishDir>
    <PublishProtocol>FileSystem</PublishProtocol>
    <TargetFramework>net5.0-windows</TargetFramework>
    <RuntimeIdentifier>win-x64</RuntimeIdentifier>
    <SelfContained>true</SelfContained>
    <PublishSingleFile>False</PublishSingleFile>
    <PublishReadyToRun>False</PublishReadyToRun>
    <PublishTrimmed>True</PublishTrimmed>
  </PropertyGroup>
</Project>

>  dotnet publish /p:PublishProfile=.\Properties\PublishProfiles\FolderProfile.xml
Microsoft (R) Build Engine version 16.11.0+0538acc04 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restored ...\winformstest\winformstest.csproj (in 87 ms).
  winformstest -> ...\winformstest\bin\Debug\net5.0-windows\win-x64\winformstest.dll
  Optimizing assemblies for size, which may change the behavior of the app. Be sure to test after publishing. See: https://aka.ms/dotnet-illink
  winformstest -> ...\winformstest\bin\Release\net5.0-windows\publish\
```

Size of `bin\Debug\net5.0-windows\win-x64\publish` is ~87MB, 204 items.

Size of `bin\Release\net5.0-windows\publixh` is ~87MB, 204 items.

I tried publishing the profile both from command line and VS, same result.

Finaly trying the same with .NET 6.0 RC1 SDK:

```
>  del .\global.json

❯  dotnet --version
6.0.100-rc.1.21430.44

❯  dotnet publish /p:PublishProfile=.\Properties\PublishProfiles\FolderProfile.xml
Microsoft (R) Build Engine version 17.0.0-preview-21430-03+e65d1aeab for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restored ...\winformstest\winformstest.csproj (in 152 ms).
C:\Program Files\dotnet\sdk\6.0.100-rc.1.21430.44\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(176,5): error NETSDK1175: Windows Forms is not supported or recommended with trimming enabled. Please go to https://aka.ms/dotnet-illink/windows-forms for more details. [...\winformstest\winformstest.csproj]

```

You're correct that we're not going to change .NET 6 to behave like .NET 5 - it worked in some limited cases but broke the app frequently and with really hard to diagnose issues. Since we din't adapt WinForms to trimming in .NET 6, we prefer to simply fail in the case, because we know that trimming .NET 6 WinForms breaks apps a lot.

Our recommendation is to:
* Don't rely on trimming WinForms - in .NET 5 it was a preview feature, in .NET 6 it fails, in either case it's not supported.
* If you have to and accept the fact that you're running an unsupported configuration, make your project stick to .NET 5 SDK using the `global.json`.

If/When we support trimming WinForms apps I'm pretty sure we will produce application size comparable to .NET 5 (and hopefully a lot better).

**Comment by Envitec** [comment]
Target Runtime: '**win-x86**'

When I publish a project from Visual Studio 2022 --> Publish Option the publish works correctly.

If I use msbuild -m /p:PublishProfile= options the publish fails with error '**obj\project.assets.json' doesn't have a target for 'net6.0/win-x86'**'

Am I missing something?

Thanks
Rajesh

**Comment by Envitec** [comment]
> @Envitec could you please create a new issue, preferably in the dotnet/sdk repo?

Done!
